### PR TITLE
Feature/override randoms

### PR DESCRIPTION
**Pull request by Shadowrs**
Allows randoms to act like APIs, can be loaded from URL in a server config rather than only being available from 1 BDN URL

### CODE REVIEW

File: src/main/java/org/parabot/core/desc/ServerProviderInfo.java [hunk 1]
```diff
@@ -50,7 +50,16 @@ public ServerProviderInfo(URL providerInfo, String username, String password) {
...
+    /**
+     * Initialize configuration with data provided by {@link org.parabot.core.parsers.servers.LocalServers} from a {@code /parabot/servers/config.json} file. Also loads the default Settings map from the BDN.
+     * @param clientJar Name of the client jar file
+     * @param hooks Name of the hooks file
+     * @param name Server name
+     * @param clientClass Entry class within the client jar
+     * @param bankTabs Bank tabs - only relevant for certain servers. Default 0
+     * @param randoms
+     */
+    public ServerProviderInfo(String clientJar, String hooks, String name, String clientClass, int bankTabs, String randoms) {
```
Comment (JKetelaar) [inline src/main/java/org/parabot/core/desc/ServerProviderInfo.java]: This is not backward compatible. Could you instead create a "new" constructor with the old parameters that basically calls this constructor?

File: src/main/java/org/parabot/core/desc/ServerProviderInfo.java [hunk 2]
```diff
@@ -141,4 +151,17 @@ public Properties getProperties() {
...
         return settings;
     }
+
+    public URL getRandoms() {
+        try {
+            String randomsUrl = properties.getProperty("randoms_jar");
+            if (randomsUrl == null || randomsUrl.length() == 0) {
+                randomsUrl = Configuration.GET_RANDOMS + (Configuration.BOT_VERSION.isNightly() ? Configuration.NIGHTLY_APPEND : "");
+            }
+            return new URL(randomsUrl);
```
Comment (JKetelaar) [inline src/main/java/org/parabot/core/desc/ServerProviderInfo.java]: Shouldn't this be **within** the if-statement, so it returns null instead?

File: src/main/java/org/parabot/core/parsers/randoms/PublicRandoms.java [hunk 3]
```diff
@@ -23,20 +23,29 @@
...
     @Override
     public void parse() {
-        File myJar = new File(Directories.getCachePath() + File.separator + fileName);
-        if (!myJar.exists() || !myJar.canRead()) {
-            download();
+
+        final File destination = new File(Directories.getCachePath() + File.separator + fileName);
+        final URL overrideDownload = Context.getInstance().getServerProviderInfo().getRandoms();
+        if (overrideDownload == null) {
+            throw new NullPointerException("randoms() URL is null - expected to default to BDN URL!");
```
Comment (JKetelaar) [inline src/main/java/org/parabot/core/parsers/randoms/PublicRandoms.java]: So you are required to fill in the randoms url?

File: src/main/java/org/parabot/core/desc/ServerProviderInfo.java (hunk 2 above)
Comment (Shadowrs) [reply src/main/java/org/parabot/core/desc/ServerProviderInfo.java]: No, if the initial `properties.getProperty("randoms_jar");` is null then it'll fall back to the default BDN URL. I'll add a comment clarifying this.

File: src/main/java/org/parabot/core/parsers/randoms/PublicRandoms.java (hunk 3 above)
Comment (Shadowrs) [reply src/main/java/org/parabot/core/parsers/randoms/PublicRandoms.java]: `   final URL overrideDownload = Context.getInstance().getServerProviderInfo().getRandoms();` will absolutely never be null, this just gives the NPE a message

Comment (JKetelaar) [reply src/main/java/org/parabot/core/parsers/randoms/PublicRandoms.java]: If it's absolutely never null, why would you add a nullcheck?

Comment (Shadowrs) [reply src/main/java/org/parabot/core/parsers/randoms/PublicRandoms.java]: I'll replace it with a comment. Shame we can't use @ NonNull :p

Comment (Shadowrs) [reply src/main/java/org/parabot/core/parsers/randoms/PublicRandoms.java]: Actually, I'll leave this exception in with an updated message. The URL will be null if the BDN URL ever changes, which of course it shouldn't, but there's no issue with some safety and a clarifying exception message.

**Pull request merged**